File: src/modules/onepiecepower.ts
import { SearchResult, Manga, Chapter } from '../classes/interfaces'
import Module from '../classes/Module'
import axios from 'axios'
import cheerio from 'cheerio'
import { NodeVM } from 'vm2'

class Juinjutsu implements Module {
  id = 'opp'
  name = 'One Piece Power'

  search(query: string) {
    return new Promise(async (resolve: (value: SearchResult[]) => void) => {
      const page = await axios.get('http://onepiecepower.info/manga/lista-manga')
      const parsed = cheerio.load(page.data)

      const res: SearchResult[] = []
      await Promise.all(parsed('body > table > tbody > tr > td').find('a').each((i, e) => {
        const el = cheerio(e)
        if(el.text().toLowerCase().includes(query.toLowerCase())) {
          res.push({
            id: el.attr('href')+'',
            title: el.text()
          })
        }
      }))

      resolve(res)
    })
  }

  manga(mangaid: string) {
    return new Promise(async (resolve: (value: Manga) => void) => {
      const page = await axios.get('http://onepiecepower.info/manga/'+mangaid)
      const parsed = cheerio.load(page.data)

      const chapters: Chapter[] = []
      await Promise.all(parsed('body > table > tbody > tr:last-child > td').find('a').each((i, e) => {
        const el = cheerio(e)
        chapters.push({
          title: el.text(),
          id: el.attr('href')+'',
          date: new Date(0)
        })
      }))

      let aut = '', syn = ''
      parsed('body > table > tbody > tr:nth-child(3) > td').find('span').each((i, e) => {
        const el = parsed(e)
        if(el.text() === 'Autore:') {
          aut = el.next('em').text()
        }
        if(el.text() === 'Descrizione:') {
          syn = el.next('em').text()
        }
      })

      const img = new URL(
        'images/cover.jpg',
        "http://onepiecepower.info/manga/"+mangaid
      ).href

      resolve({
        synopsis: syn,
        author: aut,
        artist: '',
        img: img,
        chapters: chapters.reverse(),
        sourceurl: 'http://onepiecepower.info/manga/'+mangaid
      })
    })
  }

  chapter(manga: string, id: string) {
    return new Promise(async (resolve: (value: string[]) => void) => {
      const url = new URL(id, 'http://onepiecepower.info/manga/'+manga).href
      const html = (await axios.get(url)).data

      const baseUrl = await this.pageBaseUrl(html, manga, id)

      /** calculate oversize */

      let oversize = 15
      while (true) {
        if(await this.pageExist(baseUrl, oversize+1)) {
          oversize *= 2
        } else {
          break
        }
      }

      /** bynary search */

      let start = 1;
      let end = oversize;
      let size = oversize

      while (start <= end) {
        let middle = Math.floor((start + end) / 2);
  
        if (await this.pageExist(baseUrl, middle)) {
          if(end-start <= 1) {
            size = middle
            break
          }
          start = middle + 1
        } else {
          if(end-start <= 1) {
            size = middle-1
            break
          }
          end = middle - 1
        }
      }

      /** return results */

      let res: string[] = []
      for (let i = 1; i <= size; i++) {
        res.push(this.pageUrl(baseUrl, i))
      }

      resolve(res)

    })
  }

  private pageBaseUrl(html: string, manga: string, chapter: string) {
    return new Promise<string>((resolve) => {
      let url = new URL(chapter, 'http://onepiecepower.info/manga/'+manga).href
  
      let code =
        (
          html.split('<script type="text/javascript">').pop() || ''
        )
        .split('link=link.concat(".jpg")')[0]
        .split('\n')
        .filter(v => 
          !(
            v.trim().startsWith('$') ||
            v.trim().startsWith('if') ||
            v.trim().startsWith('}') ||
            v.includes('XMLHttpRequest')
          )
        )
        .join('\n')
        .split('window.location.href')
        .join("'"+addslashes(url)+"'")
        .split('location.pathname')
        .join("'"+addslashes(new URL(url).pathname)+"'")
        .split('location.search')
        .join("''")
  
      code = code.concat('\nmodule.exports = link;')

      const s = new NodeVM()
      const res = new URL('./', (s.run(code)+'')).href
      // console.log(res)
      resolve(res)
    })
  }

  private pageUrl(baseurl: string, page: number) {
    let link = baseurl
    if(page<10){
      link=link.concat("0");
    }
    link=link.concat(page+'');
    link=link.concat(".jpg");

    return link
  }

  private async pageExist(baseurl: string, page: number) {
    return (
      (
        await axios.get(
          this.pageUrl(baseurl, page),
          {validateStatus: () => true}
        )
      ).status === 200
    )
  }

}

export default new Juinjutsu()

function addslashes( str: string ) {
  return (str + '').replace(/[\\"']/g, '\\$&').replace(/\u0000/g, '\\0');
}

File: pkg/modules/onepiecepower/onepiecepower.go
package onepiecepower

import (
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"

	"github.com/Minettyx/FoolslideProxy/pkg/types"
	"github.com/Minettyx/FoolslideProxy/pkg/utils"

	"github.com/PuerkitoBio/goquery"
	"github.com/dop251/goja"
)

type onePiecePower struct {
	baseUrl string
}

var Module = onePiecePower{
	baseUrl: "https://onepiecepower.com/",
}

var _ types.Module = onePiecePower{}

func (c onePiecePower) Id() string {
	return "opp"
}
func (c onePiecePower) Name() string {
	return "One Piece Power"
}
func (c onePiecePower) Flags() types.ModuleFlags {
	return []types.ModuleFlag{}
}

func (c onePiecePower) Search(query string) ([]types.SearchResult, error) {
	res, err := http.Get(c.baseUrl + "manga8/lista-manga")
	if err != nil {
		return nil, err
	}

	defer res.Body.Close()

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return nil, err
	}

	results := []types.SearchResult{}

	doc.Find("#allList > a").Each(func(i int, s *goquery.Selection) {
		idd, _ := s.Attr("href")

		pts := strings.Split(idd, "/")
		pts[len(pts)-1] = "images/cover.jpg"
		image := c.baseUrl + "manga8/" + strings.Join(pts, "/")

		// println(image)

		if utils.StrConaintsIgnoreCase(s.Text(), query) {
			results = append(results, types.SearchResult{
				Id:    idd,
				Title: s.Text(),
				Image: image,
			})
		}
	})

	return results, nil
}

func (c onePiecePower) Manga(id string) (*types.Manga, error) {
	res, err := http.Get(c.baseUrl + "manga8/" + id)
	if err != nil {
		return nil, err
	}

	defer res.Body.Close()

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return nil, err
	}

	key := ""
	author := ""
	artist := ""
	desc := ""
	doc.Find("body > table > tbody > tr:nth-child(3) > td > *").Each(func(i int, s *goquery.Selection) {
		if s.Is("span") {
			key = s.Text()
		} else if s.Is("em") {
			if key == "Autore:" {
				author = s.Text()
			} else if key == "Descrizione:" {
				desc = s.Text()
			} else if key == "Artista:" {
				artist = s.Text()
			}
		}
	})

	pts := strings.Split(id, "/")
	pts[len(pts)-1] = "images/cover.jpg"
	image := c.baseUrl + "manga8/" + strings.Join(pts, "/")

	chapters := []types.Chapter{}
	doc.Find("tbody > tr:nth-child(5) > td > a").Each(func(i int, s *goquery.Selection) {
		chid, _ := s.Attr("href")
		chname := s.Text()

		if strings.Contains(chname, "(Disponibile") || strings.Contains(chname, "(Available") {
			return
		}

		chapters = append(chapters, types.Chapter{
			Title: chname,
			Id:    chid,
		})
	})

	for i, j := 0, len(chapters)-1; i < j; i, j = i+1, j-1 {
		chapters[i], chapters[j] = chapters[j], chapters[i]
	}

	return &types.Manga{
		Synopsis:  desc,
		Author:    author,
		Artist:    artist,
		Img:       image,
		Sourceurl: c.baseUrl + "manga8/" + id,
		Chapters:  chapters,
	}, nil
}

func (c onePiecePower) Chapter(manga, id string) ([]string, error) {
	pts := strings.Split(manga, "/")
	pts[len(pts)-1] = id
	urls := c.baseUrl + "manga8/" + strings.Join(pts, "/")
	// urlobj, err := url.Parse(urls)
	// if err != nil {
	// 	return nil, err
	// }

	resp, err := http.Get(urls)
	if err != nil {
		return nil, err
	}

	defer resp.Body.Close()

	html, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	base, err := pageBaseUrl(string(html), manga, id)

	if err != nil {
		return nil, err
	}

	// println(base)

	// calculate capacity
	capacity := 20
	for {
		exis, err := pageExist(base, capacity+1)
		if err != nil {
			return nil, err
		}

		if exis {
			capacity *= 2
		} else {
			break
		}
	}

	// println("capacity: ", capacity)

	// binary search
	start := 1
	end := capacity
	size := capacity

	for start <= end {
		middle := (start + end) / 2
		// println("middle:", middle)

		exis, err := pageExist(base, middle)
		if err != nil {
			return nil, err
		}

		if exis {
			if end-start <= 1 {
				size = middle
				break
			}
			start = middle + 1
		} else {
			if end-start <= 1 {
				size = middle - 1
				break
			}
			end = middle - 1
		}
	}

	// println("capacity: ", size)

	results := []string{}
	for i := 1; i <= size; i++ {
		results = append(results, pageUrl(base, i))
	}

	return results, nil
}

func (c onePiecePower) Latest() ([]types.LatestResult, error) {
	return []types.LatestResult{}, nil
}

func (c onePiecePower) Popular() ([]types.PopularResult, error) {
	return []types.PopularResult{}, nil
}

func pageBaseUrl(html string, manga string, chapter string) (string, error) {
	pts := strings.Split(manga, "/")
	pts[len(pts)-1] = chapter
	urls := Module.baseUrl + "manga8/" + strings.Join(pts, "/")
	urlobj, err := url.Parse(urls)
	if err != nil {
		return "", err
	}

	code0, err := utils.StrBetween(html, `<script type="text/javascript">`, `link.concat(".jpg")`)
	if err != nil {
		return "", err
	}

	codelns := strings.Split(code0, "\n")
	codelns2 := []string{}
	for i := range codelns {
		if i == len(codelns)-1 {
			continue
		}

		p := strings.TrimSpace(codelns[i])

		if strings.HasPrefix(p, "$") || strings.Contains(p, "XMLHttpRequest") {
			continue
		}

		if strings.Contains(p, "window.location.href") {
			p = strings.ReplaceAll(p, "window.location.href", `"`+urls+`"`)
		}
		if strings.Contains(p, "location.pathname") {
			p = strings.ReplaceAll(p, "location.pathname", `"`+urlobj.Path+`"`)
		}
		if strings.Contains(p, "location.search") {
			p = strings.ReplaceAll(p, "location.search", `""`)
		}

		codelns2 = append(codelns2, p)
	}

	code := strings.Join(codelns2, "\n")

	// println(code)

	vm := goja.New()
	_, err = vm.RunString(code)
	if err != nil {
		return "", err
	}

	var res string
	err = vm.ExportTo(vm.Get("link"), &res)
	if err != nil {
		return "", err
	}

	a := strings.Split(res, "/")
	res = strings.Join(a[:len(a)-1], "/")

	// println("urls:", urls, "res: ", res)
	resobj, err := url.Parse(res)
	if err != nil {
		return "", err
	}
	if resobj.IsAbs() {
		return res + "/", nil
	} else {
		rr := urlobj.JoinPath("../", res)
		return rr.String() + "/", nil
	}
}

func pageUrl(base string, page int) string {
	if page < 10 {
		base += "0"
	}

	base += fmt.Sprint(page)
	base += ".jpg"

	return base
}

func pageExist(base string, page int) (bool, error) {
	res, err := http.Get(pageUrl(base, page))
	if err != nil {
		return false, err
	}
	return res.StatusCode == 200, nil
}
